src/wasm/wasm-stack.cpp

```diff
@@ -1656,102 +1657,36 @@ void BinaryInstWriter::emitUnreachable() {
 
 void BinaryInstWriter::mapLocalsAndEmitHeader() {
   assert(func && "BinaryInstWriter: function is not set");
-  // Map them
+  // Map params
   for (Index i = 0; i < func->getNumParams(); i++) {
     size_t curr = mappedLocals.size();
-    mappedLocals[i] = curr;
+    mappedLocals[std::make_pair(i, 0)] = curr;
   }
   for (auto type : func->vars) {
-    numLocalsByType[type]++;
+    for (auto t : type.expand()) {
+      numLocalsByType[t]++;
+    }
   }
   std::map<Type, size_t> currLocalsByType;
   for (Index i = func->getVarIndexBase(); i < func->getNumLocals(); i++) {
-    size_t index = func->getVarIndexBase();
-    Type type = func->getLocalType(i);
-    // increment now for simplicity, must decrement it in returns
-    currLocalsByType[type]++;
-    if (type == Type::i32) {
-      mappedLocals[i] = index + currLocalsByType[Type::i32] - 1;
-      continue;
-    }
-    index += numLocalsByType[Type::i32];
-    if (type == Type::i64) {
-      mappedLocals[i] = index + currLocalsByType[Type::i64] - 1;
-      continue;
-    }
-    index += numLocalsByType[Type::i64];
-    if (type == Type::f32) {
-      mappedLocals[i] = index + currLocalsByType[Type::f32] - 1;
-      continue;
-    }
-    index += numLocalsByType[Type::f32];
-    if (type == Type::f64) {
-      mappedLocals[i] = index + currLocalsByType[Type::f64] - 1;
-      continue;
-    }
-    index += numLocalsByType[Type::f64];
-    if (type == Type::v128) {
-      mappedLocals[i] = index + currLocalsByType[Type::v128] - 1;
-      continue;
-    }
-    index += numLocalsByType[Type::v128];
-    if (type == Type::funcref) {
-      mappedLocals[i] = index + currLocalsByType[Type::funcref] - 1;
-      continue;
-    }
-    index += numLocalsByType[Type::funcref];
-    if (type == Type::anyref) {
-      mappedLocals[i] = index + currLocalsByType[Type::anyref] - 1;
-      continue;
-    }
-    index += numLocalsByType[Type::anyref];
-    if (type == Type::nullref) {
-      mappedLocals[i] = index + currLocalsByType[Type::nullref] - 1;
-      continue;
-    }
-    index += numLocalsByType[Type::nullref];
-    if (type == Type::exnref) {
-      mappedLocals[i] = index + currLocalsByType[Type::exnref] - 1;
-      continue;
+    const std::vector<Type> types = func->getLocalType(i).expand();
+    for (Index j = 0; j < types.size(); j++) {
+      Type type = types[j];
+      auto varIndex = std::make_pair(i, j);
```

Comment:
the name `varIndex` is maybe confusing as it appears inside `getVarIndexBase` for example (just a few lines above) but is not directly related. Perhaps `fullIndex` or `varTupleIndex`?